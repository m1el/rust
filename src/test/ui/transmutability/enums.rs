#![feature(transmutability)]
#![feature(marker_trait_attr)]
#![allow(dead_code)]
#![allow(incomplete_features)]
#![allow(conflicting_repr_hints)]

fn test_well_defined() {
    #[repr(C, u8)]
    pub enum WellDefined { A }
    pub enum Unspecified { A }

    assert_is_transmutable_assume_nothing::<WellDefined, WellDefined>();
    assert_is_transmutable_assume_nothing::<WellDefined, Unspecified>(); //~ ERROR not satisfied
    assert_is_transmutable_assume_nothing::<Unspecified, WellDefined>(); //~ ERROR not satisfied
    assert_is_transmutable_assume_nothing::<Unspecified, Unspecified>(); //~ ERROR not satisfied
}

fn test_read_tag() {
    #[repr(C, u8)]
    pub enum OnlyTag { A }

    assert_is_transmutable_assume_nothing::<OnlyTag, u8>();
    // cannot write to tag without verifying
    assert_is_transmutable_assume_nothing::<u8, OnlyTag>(); //~ ERROR not satisfied
    // can write to tag if validity is checked
    assert_is_transmutable_assume_validity::<u8, OnlyTag>();
}

fn test_with_data() {
    #[repr(C, u8)]
    pub enum TwoU8 { A(u8), B(u8) }
    #[repr(C)]
    pub struct EquivalentStruct { tag: bool, value: u8 }
    assert_is_transmutable_assume_nothing::<TwoU8, EquivalentStruct>();
    assert_is_transmutable_assume_nothing::<EquivalentStruct, TwoU8>();
}

fn test_fail_with_data() {
    #[repr(C, u8)]
    pub enum TwoU8 { A(u8), B(bool) }
    #[repr(C)]
    pub struct Same { tag: bool, value: u8 }
    assert_is_transmutable_assume_nothing::<TwoU8, Same>();
    assert_is_transmutable_assume_nothing::<Same, TwoU8>(); //~ ERROR not satisfied
}

fn test_cursed_enum() {
    #[repr(C, u8)] enum EnumA { A(bool), B(u8), }
    #[repr(C)] struct StructA { a: bool, b: EnumA }

    #[repr(C, u8)] enum EnumB { A(bool), B(bool) }
    #[repr(C)] struct StructB { a: EnumB, b: bool }
    assert_is_transmutable_assume_nothing::<StructA, StructB>(); //~ ERROR not satisfied
    assert_is_transmutable_assume_nothing::<StructB, StructA>();
}

use std::mem::BikeshedIntrinsicFrom;

struct Context;

fn assert_is_transmutable_assume_nothing<Src, Dst>()
where
    Dst:
        BikeshedIntrinsicFrom<Src, Context, false, false, false, false>
{}

fn assert_is_transmutable_assume_validity<Src, Dst>()
where
    Dst:
        BikeshedIntrinsicFrom<Src, Context, false, false, true, false>
{}

fn main() {}
