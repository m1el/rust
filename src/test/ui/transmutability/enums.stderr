error[E0277]: the trait bound `Unspecified: BikeshedIntrinsicFrom<WellDefined, Context, false, false, false, false>` is not satisfied
  --> $DIR/enums.rs:13:58
   |
LL |     assert_is_transmutable_assume_nothing::<WellDefined, Unspecified>();
   |                                                          ^^^^^^^^^^^ the trait `BikeshedIntrinsicFrom<WellDefined, Context, false, false, false, false>` is not implemented for `Unspecified`
   |
note: required by a bound in `assert_is_transmutable_assume_nothing`
  --> $DIR/enums.rs:65:9
   |
LL | fn assert_is_transmutable_assume_nothing<Src, Dst>()
   |    ------------------------------------- required by a bound in this
...
LL |         BikeshedIntrinsicFrom<Src, Context, false, false, false, false>
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `assert_is_transmutable_assume_nothing`

error[E0277]: the trait bound `WellDefined: BikeshedIntrinsicFrom<Unspecified, Context, false, false, false, false>` is not satisfied
  --> $DIR/enums.rs:14:58
   |
LL |     assert_is_transmutable_assume_nothing::<Unspecified, WellDefined>();
   |                                                          ^^^^^^^^^^^ the trait `BikeshedIntrinsicFrom<Unspecified, Context, false, false, false, false>` is not implemented for `WellDefined`
   |
note: required by a bound in `assert_is_transmutable_assume_nothing`
  --> $DIR/enums.rs:65:9
   |
LL | fn assert_is_transmutable_assume_nothing<Src, Dst>()
   |    ------------------------------------- required by a bound in this
...
LL |         BikeshedIntrinsicFrom<Src, Context, false, false, false, false>
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `assert_is_transmutable_assume_nothing`

error[E0277]: the trait bound `Unspecified: BikeshedIntrinsicFrom<Unspecified, Context, false, false, false, false>` is not satisfied
  --> $DIR/enums.rs:15:58
   |
LL |     assert_is_transmutable_assume_nothing::<Unspecified, Unspecified>();
   |                                                          ^^^^^^^^^^^ the trait `BikeshedIntrinsicFrom<Unspecified, Context, false, false, false, false>` is not implemented for `Unspecified`
   |
note: required by a bound in `assert_is_transmutable_assume_nothing`
  --> $DIR/enums.rs:65:9
   |
LL | fn assert_is_transmutable_assume_nothing<Src, Dst>()
   |    ------------------------------------- required by a bound in this
...
LL |         BikeshedIntrinsicFrom<Src, Context, false, false, false, false>
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `assert_is_transmutable_assume_nothing`

error[E0277]: the trait bound `FakeBool: BikeshedIntrinsicFrom<u8, Context, false, false, false, false>` is not satisfied
  --> $DIR/enums.rs:25:49
   |
LL |     assert_is_transmutable_assume_nothing::<u8, FakeBool>();
   |                                                 ^^^^^^^^ the trait `BikeshedIntrinsicFrom<u8, Context, false, false, false, false>` is not implemented for `FakeBool`
   |
note: required by a bound in `assert_is_transmutable_assume_nothing`
  --> $DIR/enums.rs:65:9
   |
LL | fn assert_is_transmutable_assume_nothing<Src, Dst>()
   |    ------------------------------------- required by a bound in this
...
LL |         BikeshedIntrinsicFrom<Src, Context, false, false, false, false>
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `assert_is_transmutable_assume_nothing`

error[E0277]: the trait bound `test_fail_with_data::TwoU8: BikeshedIntrinsicFrom<Same, Context, false, false, false, false>` is not satisfied
  --> $DIR/enums.rs:45:51
   |
LL |     assert_is_transmutable_assume_nothing::<Same, TwoU8>();
   |                                                   ^^^^^ the trait `BikeshedIntrinsicFrom<Same, Context, false, false, false, false>` is not implemented for `test_fail_with_data::TwoU8`
   |
note: required by a bound in `assert_is_transmutable_assume_nothing`
  --> $DIR/enums.rs:65:9
   |
LL | fn assert_is_transmutable_assume_nothing<Src, Dst>()
   |    ------------------------------------- required by a bound in this
...
LL |         BikeshedIntrinsicFrom<Src, Context, false, false, false, false>
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `assert_is_transmutable_assume_nothing`

error[E0277]: the trait bound `StructB: BikeshedIntrinsicFrom<StructA, Context, false, false, false, false>` is not satisfied
  --> $DIR/enums.rs:54:54
   |
LL |     assert_is_transmutable_assume_nothing::<StructA, StructB>();
   |                                                      ^^^^^^^ the trait `BikeshedIntrinsicFrom<StructA, Context, false, false, false, false>` is not implemented for `StructB`
   |
note: required by a bound in `assert_is_transmutable_assume_nothing`
  --> $DIR/enums.rs:65:9
   |
LL | fn assert_is_transmutable_assume_nothing<Src, Dst>()
   |    ------------------------------------- required by a bound in this
...
LL |         BikeshedIntrinsicFrom<Src, Context, false, false, false, false>
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `assert_is_transmutable_assume_nothing`

error: aborting due to 6 previous errors

For more information about this error, try `rustc --explain E0277`.
